ports/esp32/machine_i2s.c

```diff
@@ -0,0 +1,792 @@
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_pos_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    //
+    // ---- Check validity of arguments ----
+    //
+
+    // are Pins valid?
+    int8_t sck = args[ARG_sck].u_obj == MP_OBJ_NULL ? -1 : mp_hal_get_pin_obj(args[ARG_sck].u_obj);
+    int8_t ws = args[ARG_ws].u_obj == MP_OBJ_NULL ? -1 : mp_hal_get_pin_obj(args[ARG_ws].u_obj);
+    int8_t sd = args[ARG_sd].u_obj == MP_OBJ_NULL ? -1 : mp_hal_get_pin_obj(args[ARG_sd].u_obj);
+
+    // is Mode valid?
+    i2s_mode_t mode = args[ARG_mode].u_int;
+    if ((mode != (I2S_MODE_MASTER | I2S_MODE_RX)) &&
+        (mode != (I2S_MODE_MASTER | I2S_MODE_TX))) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid mode"));
+    }
+
+    // is Bits valid?
+    i2s_bits_per_sample_t bits = args[ARG_bits].u_int;
+    if ((bits != I2S_BITS_PER_SAMPLE_16BIT) &&
+        (bits != I2S_BITS_PER_SAMPLE_32BIT)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid bits"));
+    }
+
+    // is Format valid?
+    format_t format = args[ARG_format].u_int;
+    if ((format != STEREO) &&
+        (format != MONO)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid format"));
+    }
+
+    // is Rate valid?
+    // Not checked:  ESP-IDF I2S API does not indicate a valid range for sample rate
+
+    // is Bufferlen valid?
+    // Not checked: ESP-IDF I2S API will return error if requested buffer size exceeds available memory
+
+    self->sck = sck;
+    self->ws = ws;
+    self->sd = sd;
+    self->mode = mode;
+    self->bits = bits;
+    self->format = format;
+    self->rate = args[ARG_rate].u_int;
+    self->bufferlen = args[ARG_bufferlen].u_int;
+    self->callback_for_non_blocking = MP_OBJ_NULL;
+    self->i2s_event_queue = NULL;
+    self->non_blocking_mode_queue = NULL;
+    self->non_blocking_mode_task = NULL;
+    self->io_mode = BLOCKING;
+
+    i2s_driver_uninstall(self->port);  // uninstall any previous configuration
+
+    i2s_config_t i2s_config;
+    i2s_config.communication_format = I2S_COMM_FORMAT_I2S;
+    i2s_config.mode = mode;
+    i2s_config.bits_per_sample = get_dma_bits(mode, bits);
+    i2s_config.channel_format = get_dma_format(mode, format);
+    i2s_config.sample_rate = self->rate;
+    i2s_config.intr_alloc_flags = ESP_INTR_FLAG_LOWMED;
+    i2s_config.dma_buf_count = get_dma_buf_count(mode, bits, format, self->bufferlen);
+    i2s_config.dma_buf_len = DMA_BUF_LEN_IN_I2S_FRAMES;
+    i2s_config.use_apll = false;
+
+    // I2S queue size equals the number of DMA buffers
+    check_esp_err(i2s_driver_install(self->port, &i2s_config, i2s_config.dma_buf_count, &self->i2s_event_queue));
+
+    // apply low-level workaround for bug in some ESP-IDF versions that swap
+    // the left and right channels
+    // https://github.com/espressif/esp-idf/issues/6625
+    REG_SET_BIT(I2S_CONF_REG(self->port), I2S_TX_MSB_RIGHT);
+    REG_SET_BIT(I2S_CONF_REG(self->port), I2S_RX_MSB_RIGHT);
```

Comment:
Is this still required - I think it might have been fixed on esspressif's side of things.

Reply:
This bug still exists in the versions of ESP-IDF used by MicroPython.  I have not seen any indication that Espressif has fixed the bug in new versions of the ESP-IDF.